source/userguide/custom_action.rst

```diff
@@ -0,0 +1,219 @@
+deploy conversational applications.
+
+
+Why Custom Actions?
+-------------------
+
+Suppose that a team is working to build a MindMeld application and integrate it the rest of their microservices.
+However, since the rest of the microservices are written in Java, the developers want to reuse as much logic as
+possible. Since the MindMeld application is in Python, it will be a steep curve to reimplement
+
+With Custom Actions, the developers can now shift the responsibility of fulfilling business logic to services that
+are outside of MindMeld applications, and these services can be implemented in any language that the developers
+desire. The developers can specify the exact conditions (for example, matching a certain domain or intent) to execute
+the custom actions and the MindMeld application will interact with the custom action servers through http requests.
+
+
+OpenAPI Specification for client-server interaction
+---------------------------------------------------
+
+Our client – server protocol is documented `here <https://github.com/cisco/mindmeld/blob/master/mindmeld/openapi/custom_action.yaml>`_.
+For easy viewing you can check out Swagger UI, which can also help to render OpenAPI protocol into client and server
+stubs in a variety of languages.
+
+In a broad view, each request to the Custom Action Server includes two fields: ``request`` and ``responder``. The
+``request`` field encapsulates information from the ``Request`` object (text, NLP information, etc.) and the
+``responder`` field the ``Responder`` object (directives, frame, slots, params).
+
+In a normal MindMeld application each object is passed into the application handler for processing.
+
+.. code-block:: python
+
+    @app.handle(intent='some_intent')
+    def handle_intent(request, responder):
+        # normal MindMeld application logic goes here
+
+In an analogous process, when we invoke the custom action, we are serializing the ``request`` and ``responder`` and
+passing them into the body of the http request to the server. In the response, we simply expect it to contain all
+the fields of the `Responder` object which will be deserialized and processed before returning to the end client.
+
+
+Configure Custom Actions for your MindMeld application
+------------------------------------------------------
+
+You can specify custom action configuration in ``config.py`` with the ``url`` field:
+
+.. code-block:: python
+
+    CUSTOM_ACTION_CONFIG = {"url": "http://0.0.0.0:8080/"}
+
+Currently MindMeld also supports SSL encryption by specifying the following fields in
+the ``CUSTOM_ACTION_CONFIG``: ``cert``, ``public_key`` and ``private_key``. Each field
+is the path to the local location of the certificate, public and private key.
+
+
+A Simple Custom Action Server
+-----------------------------
+
+To understand the behavior of Custom Actions, let's take a look at example of a Custom
```

Comment:
More context has to be provided on how to create this Server application. For example, we mention importing these libs `from ..models.data import Data` but this assumes a certain dir structure which the developer doesnt know about.